File: skill_manager/application/mcp/mutations.py
from __future__ import annotations

from dataclasses import replace
from typing import Iterable

from skill_manager.errors import MutationError

from .enrichment import McpEnrichmentService
from .install_resolver import resolve_registry_server_spec
from .marketplace.catalog import McpMarketplaceCatalog
from .planner import McpAdoptionPlanner
from .read_models import McpReadModelService
from .redaction import redacted_spec_dict
from .store import McpServerSpec, McpServerStore, McpSource


class McpMutationService:
    """Mutations for observed MCP configs.

    The managed manifest stores the canonical observed config. Harness files are
    projections of that canonical spec.
    """

    def __init__(
        self,
        *,
        store: McpServerStore,
        read_models: McpReadModelService,
        planner: McpAdoptionPlanner,
        marketplace_catalog: McpMarketplaceCatalog,
        enrichment: McpEnrichmentService | None = None,
    ) -> None:
        self.store = store
        self.read_models = read_models
        self.planner = planner
        self.marketplace = marketplace_catalog
        self.enrichment = enrichment

    # Install / uninstall ---------------------------------------------------

    def install_from_marketplace(
        self,
        qualified_name: str,
        *,
        source_harness: str,
        config: dict[str, object] | None = None,
    ) -> dict[str, object]:
        if not qualified_name:
            raise MutationError("qualifiedName is required", status=400)
        if not source_harness:
            raise MutationError("sourceHarness is required", status=400)
        self._require_install_target(source_harness)

        existing = self._managed_for_marketplace(qualified_name)
        if existing is not None:
            raise MutationError(
                f"a server named '{existing.name}' is already installed",
                status=409,
            )
        detail = self._marketplace_install_detail(qualified_name)
        if detail is None:
            raise MutationError(f"server not found in marketplace: {qualified_name}", status=404)
        source_spec = resolve_registry_server_spec(detail, config=config)
        if self.store.get_managed(source_spec.name) is not None:
            raise MutationError(
                f"a server named '{source_spec.name}' is already installed",
                status=409,
            )

        stored = self.store.upsert_from_spec(source_spec)
        adapter = self._source_adapter(source_harness)
        try:
            adapter.enable_server(stored)
        except Exception:  # noqa: BLE001
            self.store.remove(stored.name)
            self.read_models.invalidate()
            raise
        self.read_models.invalidate()
        return {"ok": True, "server": redacted_spec_dict(stored)}

    def install_targets(self) -> dict[str, object]:
        return {"targets": self._resolved_install_targets()}

    def uninstall_server(self, name: str) -> dict[str, object]:
        if self.store.get_managed(name) is None:
            raise MutationError(f"unknown server: {name}", status=404)
        bound_harnesses = self._harnesses_in_states(name, {"managed", "drifted"})
        succeeded: list[str] = []
        failures: list[dict[str, str]] = []
        for adapter in self.read_models.enabled_adapters():
            if adapter.harness not in bound_harnesses:
                continue
            try:
                adapter.disable_server(name)
                succeeded.append(adapter.harness)
            except Exception as error:  # noqa: BLE001
                failures.append({"harness": adapter.harness, "error": str(error)})
        if not failures:
            self.store.remove(name)
        if succeeded or not failures:
            self.read_models.invalidate()
        return {
            "ok": not failures,
            "succeeded": succeeded,
            "failed": failures,
        }

    # Per-harness toggle ----------------------------------------------------

    def enable_server(self, name: str, harness: str) -> dict[str, bool]:
        spec = self._require_server(name)
        adapter = self.read_models.require_enabled_adapter(harness)
        if adapter.has_binding(name):
            return {"ok": True}
        adapter.enable_server(spec)
        self.read_models.invalidate()
        return {"ok": True}

    def disable_server(self, name: str, harness: str) -> dict[str, bool]:
        if self.store.get_managed(name) is None:
            raise MutationError(f"unknown server: {name}", status=404)
        adapter = self.read_models.require_enabled_adapter(harness)
        adapter.disable_server(name)
        self.read_models.invalidate()
        return {"ok": True}

    def set_server_all_harnesses(self, name: str, target: str) -> dict[str, object]:
        if target not in ("enabled", "disabled"):
            raise MutationError("target must be 'enabled' or 'disabled'", status=400)
        spec = self._require_server(name)

        bound_now = self._harnesses_in_states(name, {"managed", "drifted"})

        succeeded: list[str] = []
        failures: list[dict[str, str]] = []
        flipped_any = False

        adapters = (
            self.read_models.enabled_writable_adapters()
            if target == "enabled"
            else self.read_models.enabled_addressable_adapters()
        )
        for adapter in adapters:
            if target == "enabled" and adapter.harness in bound_now:
                continue
            if target == "disabled" and adapter.harness not in bound_now:
                continue
            try:
                if target == "enabled":
                    adapter.enable_server(spec)
                else:
                    adapter.disable_server(name)
            except Exception as error:  # noqa: BLE001
                failures.append({"harness": adapter.harness, "error": str(error)})
                continue
            succeeded.append(adapter.harness)
            flipped_any = True

        if flipped_any:
            self.read_models.invalidate()

        return {
            "ok": not failures,
            "succeeded": succeeded,
            "failed": failures,
        }

    # Reconciliation -------------------------------------------------------

    def reconcile_server(
        self,
        name: str,
        *,
        source_kind: str,
        source_harness: str | None = None,
        harnesses: list[str] | None = None,
    ) -> dict[str, object]:
        if self.store.get_managed(name) is None:
            raise MutationError(f"unknown server: {name}", status=404)
        target_harnesses = (
            set(harnesses)
            if harnesses is not None
            else self._harnesses_in_states(name, {"managed", "drifted"}, addressable_only=True)
        )
        current = self._require_server(name)
        if source_kind == "managed":
            source_spec = current
        elif source_kind == "harness":
            if not source_harness:
                raise MutationError("sourceHarness is required when sourceKind is 'harness'", status=400)
            observed_spec = self._observed_spec(name, source_harness)
            source_spec = replace(
                observed_spec,
                name=current.name,
                display_name=current.display_name,
                source=current.source,
            )
            self.store.upsert_from_spec(source_spec)
            self.read_models.invalidate()
            source_spec = self._require_server(name)
        else:
            raise MutationError("sourceKind must be 'managed' or 'harness'", status=400)

        stored = self.store.get_public_spec(name) or source_spec
        binding_spec = self.store.get_binding_spec(name) or source_spec
        succeeded, failures = self._write_spec_to_harnesses(binding_spec, target_harnesses)
        if succeeded:
            self.read_models.invalidate()
        return {
            "ok": not failures,
            "server": redacted_spec_dict(stored),
            "succeeded": succeeded,
            "failed": failures,
        }

    # Adoption -------------------------------------------------------------

    def _apply_enrichment(self, spec: McpServerSpec) -> McpServerSpec:
        if self.enrichment is None:
            return spec
        link = self.enrichment.lookup(spec.name)
        if link is None:
            return spec
        return replace(
            spec,
            display_name=link.display_name or spec.display_name,
            source=McpSource.marketplace(link.qualified_name),
        )

    def adopt(
        self,
        name: str,
        *,
        source_harness: str | None = None,
        harnesses: list[str] | None = None,
    ) -> dict[str, object]:
        if self.store.get_managed(name) is not None:
            raise MutationError(
                f"a managed server named '{name}' already exists", status=409
            )
        group = self.planner.require_group(name)
        if source_harness:
            target_spec = next(
                (sighting.spec for sighting in group.sightings if sighting.harness == source_harness),
                None,
            )
            if target_spec is None:
                raise MutationError(
                    f"server '{name}' was not observed in harness '{source_harness}'",
                    status=400,
                )
        else:
            target_spec = group.canonical_spec
        if target_spec is None:
            raise MutationError(
                f"server '{name}' has different configs across harnesses; choose a sourceHarness to adopt",
                status=409,
            )
        if target_spec.name != name:
            target_spec = replace(target_spec, name=name)
        target_spec = self._apply_enrichment(target_spec)

        target_harnesses = set(harnesses) if harnesses else {s.harness for s in group.sightings}
        stored = self.store.upsert_from_spec(target_spec)
        stored_binding_spec = self.store.get_binding_spec(stored.name)
        if stored_binding_spec is None:
            raise MutationError(f"unknown server: {name}", status=404)

        succeeded, failures = self._write_spec_to_harnesses(
            stored_binding_spec,
            target_harnesses,
        )

        self.read_models.invalidate()
        response_spec = self.store.get_public_spec(stored.name) or stored_binding_spec
        return {
            "ok": not failures,
            "server": redacted_spec_dict(response_spec),
            "succeeded": succeeded,
            "failed": failures,
        }

    # Internal helpers -----------------------------------------------------

    def _resolved_install_targets(self) -> list[dict[str, object]]:
        enabled = set(self.read_models.enabled_harnesses())
        targets: list[dict[str, object]] = []
        for status in self.read_models.harness_statuses():
            supported = True
            reason = None
            if supported and status.harness not in enabled:
                supported = False
                reason = "Harness support is disabled"
            elif supported and not status.mcp_writable:
                supported = False
                reason = status.mcp_unavailable_reason or "MCP config is not writable for this harness"
            elif supported and not (status.installed or status.config_present):
                supported = False
                reason = f"{status.label} is not installed and has no MCP config file"
            targets.append(
                {
                    "harness": status.harness,
                    "label": status.label,
                    "logoKey": status.logo_key,
                    # Legacy frontend wire field; install now writes directly to the selected harness.
                    "smitheryClient": status.harness if supported else None,
                    "supported": supported,
                    "reason": reason,
                }
            )
        return targets

    def _require_install_target(self, harness: str) -> None:
        for target in self._resolved_install_targets():
            if target["harness"] != harness:
                continue
            if target["supported"]:
                return
            reason = target.get("reason")
            raise MutationError(str(reason or f"source harness is not installable: {harness}"), status=400)
        raise MutationError(f"unknown MCP source harness: {harness}", status=400)

    def _marketplace_install_detail(self, qualified_name: str):
        install_detail = getattr(self.marketplace, "install_detail", None)
        if callable(install_detail):
            detail = install_detail(qualified_name)
            if detail is not None:
                to_resolver_detail = getattr(detail, "to_resolver_detail", None)
                return to_resolver_detail() if callable(to_resolver_detail) else detail
        return self.marketplace.detail(qualified_name)

    def _source_adapter(self, harness: str):
        if harness not in self.read_models.enabled_harnesses():
            raise MutationError(f"harness support is disabled: {harness}", status=400)
        adapter = self.read_models.find_adapter(harness)
        if adapter is None:
            raise MutationError(f"unknown MCP source harness: {harness}", status=400)
        return adapter

    def _harnesses_in_states(
        self,
        name: str,
        states: Iterable[str],
        *,
        addressable_only: bool = False,
    ) -> set[str]:
        allowed_states = set(states)
        addressable = (
            {adapter.harness for adapter in self.read_models.enabled_addressable_adapters()}
            if addressable_only
            else set(self.read_models.enabled_harnesses())
        )
        snapshot = self.read_models.snapshot()
        result: set[str] = set()
        for scan in snapshot.harness_scans:
            if scan.harness not in addressable:
                continue
            for entry in scan.entries:
                if entry.name == name and entry.state in allowed_states:
                    result.add(scan.harness)
        return result

    def _observed_spec(self, name: str, harness: str) -> McpServerSpec:
        snapshot = self.read_models.snapshot()
        for scan in snapshot.harness_scans:
            if scan.harness != harness:
                continue
            for entry in scan.entries:
                if entry.name != name:
                    continue
                if entry.parsed_spec is None:
                    raise MutationError(
                        entry.parse_issue or f"unable to parse '{name}' in {harness}",
                        status=409,
                    )
                return entry.parsed_spec
        raise MutationError(f"server '{name}' was not observed in harness '{harness}'", status=404)

    def _write_spec_to_harnesses(
        self,
        spec: McpServerSpec,
        harnesses: Iterable[str],
    ) -> tuple[list[str], list[dict[str, str]]]:
        targets = set(harnesses)
        succeeded: list[str] = []
        failures: list[dict[str, str]] = []
        for adapter in self.read_models.enabled_adapters():
            if adapter.harness not in targets:
                continue
            try:
                adapter.enable_server(spec)
            except Exception as error:  # noqa: BLE001
                failures.append({"harness": adapter.harness, "error": str(error)})
                continue
            succeeded.append(adapter.harness)
        return succeeded, failures

    def _require_server(self, name: str) -> McpServerSpec:
        spec = self.store.get_binding_spec(name)
        if spec is None:
            raise MutationError(f"unknown server: {name}", status=404)
        return spec

    def _managed_for_marketplace(self, qualified_name: str) -> McpServerSpec | None:
        for server in self.store.list_managed():
            if server.source.kind == "marketplace" and server.source.locator == qualified_name:
                return server
        return None


__all__ = ["McpMutationService"]

File: src-tauri/src/mcp/mutations.rs
use std::collections::HashSet;

use serde_json::{json, Value};

use super::adapters::McpReadModelService;
use super::availability::{availability_cache_key, McpAvailabilityProbe};
use super::config_choice::observed_spec_from_scans;
use super::harness_application::{harnesses_in_states, McpHarnessApplication};
use super::planner::McpAdoptionPlanner;
use super::redaction::redacted_spec_dict;
use super::store::{prepare_managed_spec, McpServerSpec, McpServerStore, McpSource};
use crate::error::{ApiError, ApiResult};
use crate::marketplace::McpMarketplaceService;
use super::availability::McpEnrichmentService;

#[derive(Clone)]
pub struct McpMutationService {
    store: McpServerStore,
    read_models: McpReadModelService,
    planner: McpAdoptionPlanner,
    marketplace: McpMarketplaceService,
    enrichment: McpEnrichmentService,
    harness_application: McpHarnessApplication,
    availability_probe: McpAvailabilityProbe,
    availability_cache: std::sync::Arc<std::sync::Mutex<std::collections::HashMap<(String, String), super::availability::McpAvailabilityResult>>>,
}

impl McpMutationService {
    pub fn new(
        store: McpServerStore,
        read_models: McpReadModelService,
        planner: McpAdoptionPlanner,
        marketplace: McpMarketplaceService,
        enrichment: McpEnrichmentService,
    ) -> Self {
        let harness_application = McpHarnessApplication::new(read_models.clone());
        Self {
            store,
            read_models,
            planner,
            marketplace,
            enrichment,
            harness_application,
            availability_probe: McpAvailabilityProbe::default(),
            availability_cache: std::sync::Arc::new(std::sync::Mutex::new(std::collections::HashMap::new())),
        }
    }

    pub async fn install_from_marketplace(&self, qualified_name: &str) -> ApiResult<Value> {
        if qualified_name.trim().is_empty() {
            return Err(ApiError::bad_request("qualifiedName is required"));
        }
        let detail = self
            .marketplace
            .install_detail(qualified_name)
            .await?
            .ok_or_else(|| ApiError::not_found(format!("server not found in marketplace: {qualified_name}")))?;
        let mut spec = spec_from_marketplace_detail(&detail, qualified_name)?;
        let reinstalled = if let Some(existing) = self.managed_for_marketplace(qualified_name) {
            spec.name = existing.name;
            true
        } else if self.store.get_record(&spec.name).is_some() {
            return Err(ApiError::conflict(format!(
                "a server named '{}' is already installed",
                spec.name
            )));
        } else {
            false
        };
        let stored = self.store.upsert(spec).map_err(ApiError::bad_request)?;
        self.read_models.invalidate();
        if let Ok(mut cache) = self.availability_cache.lock() {
            cache.insert(
                availability_cache_key(&stored.name, &stored),
                self.availability_probe.probe(&stored),
            );
        }
        Ok(json!({
            "ok": true,
            "reinstalled": reinstalled,
            "server": redacted_spec_dict(&stored)
        }))
    }

    pub fn uninstall_server(&self, name: &str) -> ApiResult<Value> {
        if self.store.get_record(name).is_none() {
            return Err(ApiError::not_found(format!("unknown server: {name}")));
        }
        let bound = harnesses_in_states(&self.read_models, name, &["managed", "drifted"], false);
        let result = self.harness_application.disable_many(name, &bound, false);
        if result.ok() {
            self.store.remove(name).map_err(ApiError::internal)?;
        }
        Ok(result.to_json())
    }

    pub fn enable_server(&self, name: &str, harness: &str, _config: Option<Value>) -> ApiResult<Value> {
        let spec = self
            .store
            .get_record(name)
            .ok_or_else(|| ApiError::not_found(format!("unknown server: {name}")))?;
        let adapter = self
            .read_models
            .require_enabled_adapter(harness)
            .map_err(ApiError::bad_request)?;
        if adapter.has_binding(name) {
            return Ok(json!({ "ok": true }));
        }
        let result = self.harness_application.enable_one(adapter, &spec);
        if !result.failed.is_empty() {
            return Err(ApiError::bad_request(
                result.failed[0]
                    .get("error")
                    .and_then(|v| v.as_str())
                    .unwrap_or("enable failed")
                    .to_string(),
            ));
        }
        Ok(json!({ "ok": true }))
    }

    pub fn disable_server(&self, name: &str, harness: &str) -> ApiResult<Value> {
        if self.store.get_record(name).is_none() {
            return Err(ApiError::not_found(format!("unknown server: {name}")));
        }
        let adapter = self
            .read_models
            .require_enabled_adapter(harness)
            .map_err(ApiError::bad_request)?;
        adapter.disable_server(name).map_err(ApiError::bad_request)?;
        self.read_models.invalidate();
        Ok(json!({ "ok": true }))
    }

    pub fn set_server_all_harnesses(
        &self,
        name: &str,
        target: &str,
        _config: Option<Value>,
    ) -> ApiResult<Value> {
        if target != "enabled" && target != "disabled" {
            return Err(ApiError::bad_request("target must be 'enabled' or 'disabled'"));
        }
        let spec = self
            .store
            .get_record(name)
            .ok_or_else(|| ApiError::not_found(format!("unknown server: {name}")))?;
        let bound_now = harnesses_in_states(&self.read_models, name, &["managed", "drifted"], false);
        let result = if target == "enabled" {
            self.harness_application.enable_many(
                &spec,
                &self.read_models.enabled_harnesses(),
                true,
                &bound_now,
            )
        } else {
            self.harness_application
                .disable_many(name, &bound_now, true)
        };
        Ok(result.to_json())
    }

    pub fn reconcile_server(
        &self,
        name: &str,
        source_kind: &str,
        observed_harness: Option<String>,
        harnesses: Option<Vec<String>>,
    ) -> ApiResult<Value> {
        if self.store.get_record(name).is_none() {
            return Err(ApiError::not_found(format!("unknown server: {name}")));
        }
        let target_harnesses: HashSet<String> = if let Some(list) = harnesses {
            list.into_iter().collect()
        } else {
            harnesses_in_states(&self.read_models, name, &["managed", "drifted"], true)
                .into_iter()
                .collect()
        };
        let current = self
            .store
            .get_record(name)
            .ok_or_else(|| ApiError::not_found(format!("unknown server: {name}")))?;
        let source_spec = match source_kind {
            "managed" => current.clone(),
            "harness" => {
                let observed = observed_harness.ok_or_else(|| {
                    ApiError::bad_request("observedHarness is required when sourceKind is 'harness'")
                })?;
                let snapshot = self.read_models.snapshot();
                let mut observed_spec =
                    observed_spec_from_scans(name, &observed, &snapshot.harness_scans)
                        .map_err(ApiError::bad_request)?;
                observed_spec.name = current.name.clone();
                observed_spec.display_name = current.display_name.clone();
                observed_spec.source = current.source.clone();
                observed_spec
            }
            _ => return Err(ApiError::bad_request("sourceKind must be 'managed' or 'harness'")),
        };
        let result = self.harness_application.enable_many(
            &source_spec,
            &target_harnesses.into_iter().collect::<Vec<_>>(),
            false,
            &[],
        );
        if !result.succeeded.is_empty() && source_spec.revision != current.revision {
            self.store.upsert(source_spec.clone()).map_err(ApiError::bad_request)?;
        }
        let stored = self.store.get_record(name).unwrap_or(source_spec);
        let mut payload = result.to_json();
        if let Some(obj) = payload.as_object_mut() {
            obj.insert("server".into(), redacted_spec_dict(&stored));
        }
        Ok(payload)
    }

    pub fn adopt(
        &self,
        name: &str,
        observed_harness: Option<&str>,
        harnesses: Option<Vec<String>>,
    ) -> ApiResult<Value> {
        if self.store.get_record(name).is_some() {
            return Err(ApiError::conflict(format!(
                "a managed server named '{name}' already exists"
            )));
        }
        let group = self
            .planner
            .require_group(name)
            .map_err(|error| ApiError::not_found(error))?;
        let target_spec = if let Some(harness) = observed_harness {
            group
                .sightings
                .iter()
                .find(|s| s.harness == harness)
                .map(|s| s.spec.clone())
                .ok_or_else(|| {
                    ApiError::bad_request(format!(
                        "server '{name}' was not observed in harness '{harness}'"
                    ))
                })?
        } else {
            group.canonical_spec.clone().ok_or_else(|| {
                ApiError::conflict(format!(
                    "server '{name}' has different configs across harnesses; choose an observedHarness to adopt"
                ))
            })?
        };
        let mut target_spec = if target_spec.name != name {
            McpServerSpec {
                name: name.to_string(),
                ..target_spec
            }
        } else {
            target_spec
        };
        target_spec = self.apply_enrichment(target_spec);
        let target_harnesses: HashSet<String> = if let Some(list) = harnesses {
            list.into_iter().collect()
        } else {
            group.sightings.iter().map(|s| s.harness.clone()).collect()
        };
        let result = self.harness_application.enable_many(
            &target_spec,
            &target_harnesses.into_iter().collect::<Vec<_>>(),
            false,
            &[],
        );
        if !result.succeeded.is_empty() {
            self.store
                .upsert(prepare_managed_spec(target_spec.clone()))
                .map_err(ApiError::bad_request)?;
        }
        let stored = self.store.get_record(name).unwrap_or(target_spec);
        let mut payload = result.to_json();
        if let Some(obj) = payload.as_object_mut() {
            obj.insert("server".into(), redacted_spec_dict(&stored));
        }
        Ok(payload)
    }

    fn apply_enrichment(&self, mut spec: McpServerSpec) -> McpServerSpec {
        let Some(link) = self.enrichment.lookup(&spec.name) else {
            return spec;
        };
        spec.display_name = if link.display_name.is_empty() {
            spec.display_name
        } else {
            link.display_name
        };
        spec.source = McpSource::marketplace(&link.qualified_name);
        spec
    }

    fn managed_for_marketplace(&self, qualified_name: &str) -> Option<McpServerSpec> {
        self.store.list_records().into_iter().find(|server| {
            server.source.kind == "marketplace" && server.source.locator == qualified_name
        })
    }
}

fn spec_from_marketplace_detail(detail: &Value, qualified_name: &str) -> Result<McpServerSpec, ApiError> {
    let name = detail
        .get("managedName")
        .or_else(|| detail.get("qualifiedName"))
        .or_else(|| detail.get("displayName"))
        .and_then(|v| v.as_str())
        .unwrap_or(qualified_name)
        .to_string();
    let connection = detail.get("connection");
    let transport = connection
        .and_then(|v| v.get("kind"))
        .or_else(|| connection.and_then(|v| v.get("type")))
        .and_then(|v| v.as_str())
        .unwrap_or("stdio")
        .to_string();
    Ok(McpServerSpec {
        name: name.clone(),
        display_name: detail
            .get("displayName")
            .and_then(|v| v.as_str())
            .unwrap_or(&name)
            .to_string(),
        source: McpSource::marketplace(qualified_name),
        transport: transport.clone(),
        command: Some("npx".into()),
        args: Some(vec!["-y".into(), format!("{qualified_name}@latest")]),
        env: None,
        url: if transport == "http" || transport == "sse" {
            connection
                .and_then(|v| v.get("deploymentUrl"))
                .and_then(|v| v.as_str())
                .map(str::to_string)
        } else {
            None
        },
        headers: None,
        installed_at: String::new(),
        revision: String::new(),
    })
}
